File: bridges/discord/platform.ts
/**
 * Cross-platform abstractions for paths, processes, signals, environment, and scheduling.
 *
 * Phases W1+W2 of cross-platform plan.
 * All platform-specific logic in the TypeScript bot code funnels through here.
 */

import { existsSync, readdirSync } from "fs";
import { join } from "path";
import { homedir, platform, tmpdir } from "os";
import { execSync } from "child_process";

export const IS_WINDOWS = platform() === "win32";
export const IS_MACOS = platform() === "darwin";
export const IS_LINUX = platform() === "linux";

// ─── Paths ─────────────────────────────────────────────────────────────

export const paths = {
  /** User home directory. */
  home(): string {
    return process.env.HOME || process.env.USERPROFILE || homedir();
  },

  /** Path to Python 3 interpreter. */
  python(): string {
    if (process.env.PYTHON_PATH) return process.env.PYTHON_PATH;

    if (IS_MACOS) {
      const brew = "/opt/homebrew/bin/python3";
      if (existsSync(brew)) return brew;
    }

    // Fallback: assume it's in PATH
    return IS_WINDOWS ? "python" : "python3";
  },

  /** Path to Claude CLI binary. */
  claudeCli(): string {
    if (process.env.CLAUDE_CLI_PATH) return process.env.CLAUDE_CLI_PATH;

    if (IS_WINDOWS) {
      const localApp = process.env.LOCALAPPDATA || "";
      const candidates = [
        join(localApp, "claude", "claude.exe"),
        join(localApp, "Programs", "claude", "claude.exe"),
      ];
      for (const c of candidates) {
        if (existsSync(c)) return c;
      }
      return "claude"; // hope it's in PATH
    }

    return join(paths.home(), ".local", "bin", "claude");
  },

  /** Platform temp directory. */
  tempDir(): string {
    return tmpdir();
  },

  /** PATH separator (: on Unix, ; on Windows). */
  pathSeparator(): string {
    return IS_WINDOWS ? ";" : ":";
  },

  /** Build a PATH string from directory components. */
  buildPath(...dirs: string[]): string {
    return dirs.filter(Boolean).join(paths.pathSeparator());
  },

  /** Reasonable default PATH for clean subprocess environments. */
  defaultPath(): string {
    const home = paths.home();
    if (IS_WINDOWS) {
      const sysRoot = process.env.SystemRoot || "C:\\Windows";
      return paths.buildPath(
        join(home, ".local", "bin"),
        join(sysRoot, "System32"),
        sysRoot,
      );
    }
    return paths.buildPath(
      join(home, ".local", "bin"),
      "/opt/homebrew/bin",
      "/usr/local/bin",
      "/usr/bin",
      "/bin",
    );
  },

  /** Default shell for the platform. */
  defaultShell(): string {
    if (IS_WINDOWS) {
      return process.env.COMSPEC || "C:\\Windows\\System32\\cmd.exe";
    }
    return process.env.SHELL || "/bin/zsh";
  },

  /** macOS LaunchAgents directory. Null on other platforms. */
  launchAgentsDir(): string | null {
    if (IS_MACOS) return join(paths.home(), "Library", "LaunchAgents");
    return null;
  },
};

// ─── Process Management ────────────────────────────────────────────────

export const proc = {
  /** Check if a process is running. */
  isAlive(pid: number): boolean {
    try {
      if (IS_WINDOWS) {
        const result = execSync(`tasklist /FI "PID eq ${pid}" /NH`, {
          encoding: "utf-8",
          timeout: 5000,
          stdio: ["pipe", "pipe", "pipe"],
        });
        return result.includes(String(pid));
      }
      // Unix: signal 0
      process.kill(pid, 0);
      return true;
    } catch {
      return false;
    }
  },

  /**
   * Terminate a process and its descendants. Returns true if a kill signal
   * was successfully delivered (not whether anything actually died — kernel
   * delivery semantics).
   *
   * On Unix: tries process-group kill first (negative pid = group). The bot
   * spawns task runners with `detached: true` so the python child becomes
   * the group leader; group kill cascades to its `subprocess.Popen`-spawned
   * `claude` / `codex` CLI grandchild. Without group kill, signaling only
   * the leader leaves the CLI grandchild orphaned (reparented to init/launchd)
   * and continuing to spend API credits — which is exactly the bug
   * /stop hit pre-fix. Falls back to direct-pid kill if the spawn wasn't
   * detached (e.g., heartbeat scripts spawned without process-group).
   *
   * On Windows: `taskkill /F /T` already kills the tree, no change needed.
   */
  terminate(pid: number): boolean {
    try {
      if (IS_WINDOWS) {
        execSync(`taskkill /PID ${pid} /F /T`, {
          timeout: 10000,
          stdio: ["pipe", "pipe", "pipe"],
        });
        return true;
      }
      // Try group kill first; the negative-pid form signals the whole
      // process group whose leader is `pid`.
      try {
        process.kill(-pid, "SIGTERM");
        return true;
      } catch {
        // Group kill can fail with ESRCH if `pid` isn't a group leader
        // (spawn wasn't detached) or with EPERM if we don't own it. Fall
        // back to direct-pid kill — better than nothing.
        process.kill(pid, "SIGTERM");
        return true;
      }
    } catch {
      return false;
    }
  },

  /**
   * Open a URL in the default browser.
   * Uses 'open' on macOS, 'start' on Windows, 'xdg-open' on Linux.
   */
  openUrl(url: string): void {
    const cmd = IS_MACOS ? "open" : IS_WINDOWS ? "start" : "xdg-open";
    try {
      execSync(`${cmd} "${url}"`, { stdio: "ignore" });
    } catch {
      console.error(`[PLATFORM] Failed to open URL: ${url}`);
    }
  },
};

// ─── Environment ───────────────────────────────────────────────────────

export const env = {
  /**
   * Build a clean subprocess environment.
   * Strips CLAUDE* vars, sets sensible cross-platform defaults.
   */
  cleanEnv(passthrough?: string[]): Record<string, string> {
    const home = paths.home();
    const result: Record<string, string> = {
      HOME: home,
      USERPROFILE: home,
      USER: process.env.USER || process.env.USERNAME || "",
      PATH: paths.defaultPath(),
      SHELL: paths.defaultShell(),
      LANG: process.env.LANG || "en_US.UTF-8",
      TERM: "dumb",
      HARNESS_ROOT: process.env.HARNESS_ROOT || "",
    };

    if (IS_WINDOWS) {
      for (const key of ["SystemRoot", "COMSPEC", "TEMP", "TMP", "LOCALAPPDATA", "APPDATA"]) {
        if (process.env[key]) result[key] = process.env[key]!;
      }
    }

    if (passthrough) {
      for (const key of passthrough) {
        if (process.env[key]) result[key] = process.env[key]!;
      }
    }

    // Remove empty values
    for (const [k, v] of Object.entries(result)) {
      if (!v) delete result[k];
    }

    return result;
  },

  /** Return env dict with all CLAUDE* vars removed. */
  stripClaudeVars(baseEnv?: Record<string, string>): Record<string, string> {
    const source = baseEnv || { ...process.env } as Record<string, string>;
    const result: Record<string, string> = {};
    for (const [k, v] of Object.entries(source)) {
      if (!k.startsWith("CLAUDE") && v !== undefined) result[k] = v;
    }
    return result;
  },
};

// ─── Scheduler ────────────────────────────────────────────────────────

export interface TaskStatus {
  label: string;
  loaded: boolean;
  pid?: number;
  exitCode?: number;
  state: "running" | "loaded" | "not-loaded" | "stale" | "unknown";
}

export const scheduler = {
  LABEL_PREFIX: "com.aiharness.heartbeat",

  /** Name of the platform scheduler. */
  name(): string {
    if (IS_MACOS) return "launchd";
    if (IS_WINDOWS) return "task-scheduler";
    return "systemd";
  },

  /** Full label for a task. */
  taskLabel(taskName: string): string {
    return `${scheduler.LABEL_PREFIX}.${taskName}`;
  },

  /** Get status of a specific scheduled task. */
  status(label: string): TaskStatus {
    if (IS_MACOS) {
      try {
        const result = execSync(`launchctl list ${label} 2>/dev/null`, {
          encoding: "utf-8",
          timeout: 5000,
          stdio: ["pipe", "pipe", "pipe"],
        });
        let pid: number | undefined;
        for (const line of result.split("\n")) {
          if (line.includes('"PID"')) {
            const m = line.match(/=\s*(\d+)/);
            if (m) pid = parseInt(m[1], 10);
          }
        }
        return {
          label,
          loaded: true,
          pid,
          state: pid ? "running" : "loaded",
        };
      } catch {
        return { label, loaded: false, state: "not-loaded" };
      }
    }

    if (IS_WINDOWS) {
      const schtaskName = `\\${label.replace(/\./g, "\\")}`;
      try {
        const result = execSync(
          `schtasks /Query /TN "${schtaskName}" /FO CSV /V /NH 2>nul`,
          { encoding: "utf-8", timeout: 5000, stdio: ["pipe", "pipe", "pipe"] },
        );
        const running = result.includes("Running");
        return { label, loaded: true, state: running ? "running" : "loaded" };
      } catch {
        return { label, loaded: false, state: "not-loaded" };
      }
    }

    return { label, loaded: false, state: "unknown" };
  },

  /** List all tasks with the given prefix. */
  listTasks(prefix?: string): TaskStatus[] {
    const pfx = prefix || scheduler.LABEL_PREFIX;
    if (IS_MACOS) {
      try {
        const result = execSync("launchctl list 2>/dev/null", {
          encoding: "utf-8",
          timeout: 10000,
          stdio: ["pipe", "pipe", "pipe"],
        });
        const tasks: TaskStatus[] = [];
        for (const line of result.trim().split("\n")) {
          const parts = line.split("\t");
          if (parts.length < 3) continue;
          const [pidStr, exitStr, label] = [parts[0].trim(), parts[1].trim(), parts[2].trim()];
          if (!label.startsWith(pfx)) continue;
          const pid = pidStr !== "-" ? parseInt(pidStr, 10) : undefined;
          const exitCode = exitStr !== "-" ? parseInt(exitStr, 10) : undefined;
          const state = pid ? "running" : exitCode === 78 ? "stale" : "loaded";
          tasks.push({ label, loaded: true, pid, exitCode, state });
        }
        return tasks;
      } catch {
        return [];
      }
    }
    return [];
  },

  /** Reload (stop + start) a task. */
  reload(labelOrName: string): boolean {
    const label = labelOrName.includes(".") ? labelOrName : scheduler.taskLabel(labelOrName);
    if (IS_MACOS) {
      const launchDir = paths.launchAgentsDir();
      if (!launchDir) return false;
      const plistPath = join(launchDir, `${label}.plist`);
      if (!existsSync(plistPath)) return false;
      try {
        execSync(`launchctl unload "${plistPath}" 2>/dev/null`, { stdio: "pipe", timeout: 10000 });
        execSync(`launchctl load "${plistPath}"`, { stdio: "pipe", timeout: 10000 });
        return true;
      } catch {
        return false;
      }
    }
    return false;
  },
};

// ─── Signal Handling ───────────────────────────────────────────────────

/**
 * Register graceful shutdown handlers.
 * On Unix: SIGINT + SIGTERM. On Windows: SIGINT only (SIGTERM not supported).
 */
export function onShutdown(handler: () => void): void {
  process.on("SIGINT", handler);
  if (!IS_WINDOWS) {
    process.on("SIGTERM", handler);
  }
}

File: heartbeat-tasks/lib/platform.py
"""Cross-platform abstractions for paths, processes, signals, and environment.

Phase W1 of cross-platform plan.
Every platform-specific operation funnels through this module so the rest
of the codebase stays platform-agnostic.

Usage:
    from lib.platform import paths, proc, env

    paths.python()          # "/opt/homebrew/bin/python3" or "python" on Windows
    paths.claude_cli()      # "~/.local/bin/claude" or "%LOCALAPPDATA%\\claude\\claude.exe"
    paths.home()            # HOME or USERPROFILE
    paths.temp_dir()        # /tmp or %TEMP%
    proc.is_alive(pid)      # signal-0 on Unix, OpenProcess on Windows
    proc.terminate(pid)     # SIGTERM on Unix, taskkill on Windows
    env.clean_path()        # colon-separated on Unix, semicolon on Windows
"""

from __future__ import annotations

import os
import platform
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

IS_WINDOWS = platform.system() == "Windows"
IS_MACOS = platform.system() == "Darwin"
IS_LINUX = platform.system() == "Linux"


# ─── Paths ─────────────────────────────────────────────────────────────

class _Paths:
    """Platform-aware path resolution."""

    def home(self) -> str:
        """User home directory."""
        return os.environ.get("HOME") or os.environ.get("USERPROFILE") or str(Path.home())

    def python(self) -> str:
        """Path to Python 3 interpreter."""
        # Explicit env override
        configured = os.environ.get("PYTHON_PATH")
        if configured:
            return configured

        if IS_MACOS:
            # Homebrew default
            brew_python = "/opt/homebrew/bin/python3"
            if os.path.isfile(brew_python):
                return brew_python

        # Fallback: find python3 (or python on Windows) in PATH
        name = "python" if IS_WINDOWS else "python3"
        found = shutil.which(name)
        if found:
            return found

        # Windows fallback: try python3 too
        if IS_WINDOWS:
            found = shutil.which("python3")
            if found:
                return found

        return name  # last resort — hope it's in PATH

    def claude_cli(self) -> str:
        """Path to Claude CLI binary."""
        configured = os.environ.get("CLAUDE_CLI_PATH")
        if configured:
            return configured

        if IS_WINDOWS:
            # Windows: check common install locations
            local_app = os.environ.get("LOCALAPPDATA", "")
            candidates = [
                os.path.join(local_app, "claude", "claude.exe"),
                os.path.join(local_app, "Programs", "claude", "claude.exe"),
            ]
            for c in candidates:
                if os.path.isfile(c):
                    return c
            # Fallback to PATH
            found = shutil.which("claude")
            return found or "claude"

        # Unix: ~/.local/bin/claude
        return os.path.join(self.home(), ".local", "bin", "claude")

    def temp_dir(self) -> str:
        """Platform temp directory."""
        return tempfile.gettempdir()

    def path_separator(self) -> str:
        """PATH environment variable separator (: on Unix, ; on Windows)."""
        return ";" if IS_WINDOWS else ":"

    def build_path(self, *dirs: str) -> str:
        """Build a PATH string from directory components."""
        return self.path_separator().join(d for d in dirs if d)

    def default_path(self) -> str:
        """Reasonable default PATH for clean subprocess environments."""
        home = self.home()
        if IS_WINDOWS:
            system_root = os.environ.get("SystemRoot", r"C:\Windows")
            return self.build_path(
                os.path.join(home, ".local", "bin"),
                os.path.join(system_root, "System32"),
                system_root,
            )
        # macOS/Linux
        return self.build_path(
            os.path.join(home, ".local", "bin"),
            "/opt/homebrew/bin",
            "/usr/local/bin",
            "/usr/bin",
            "/bin",
        )

    def default_shell(self) -> str:
        """Default shell for the platform."""
        if IS_WINDOWS:
            return os.environ.get("COMSPEC", r"C:\Windows\System32\cmd.exe")
        return os.environ.get("SHELL", "/bin/zsh")

    def google_drive_dir(self) -> str | None:
        """Auto-detect Google Drive mount point. Returns None if not found."""
        if IS_MACOS:
            cloud_dir = os.path.expanduser("~/Library/CloudStorage")
            if os.path.isdir(cloud_dir):
                for entry in os.listdir(cloud_dir):
                    if entry.startswith("GoogleDrive-"):
                        return os.path.join(cloud_dir, entry, "My Drive")
            return None

        if IS_WINDOWS:
            # Google Drive for Desktop creates a virtual drive (G:\) or
            # streams to %USERPROFILE%\Google Drive
            candidates = [
                os.path.join(self.home(), "Google Drive"),
                os.path.join(self.home(), "My Drive"),
            ]
            # Also check drive letters G: through Z:
            for letter in "GHIJKLMNOPQRSTUVWXYZ":
                candidates.append(f"{letter}:\\My Drive")
            for c in candidates:
                if os.path.isdir(c):
                    return c
            return None

        # Linux: common locations
        for candidate in [
            os.path.expanduser("~/google-drive"),
            os.path.expanduser("~/Google Drive"),
        ]:
            if os.path.isdir(candidate):
                return candidate
        return None

    def launch_agents_dir(self) -> str | None:
        """macOS LaunchAgents directory. None on other platforms."""
        if IS_MACOS:
            return os.path.join(self.home(), "Library", "LaunchAgents")
        return None


# ─── Process Management ────────────────────────────────────────────────

class _Proc:
    """Cross-platform process management."""

    def is_alive(self, pid: int) -> bool:
        """Check if a process is running."""
        if IS_WINDOWS:
            try:
                result = subprocess.run(
                    ["tasklist", "/FI", f"PID eq {pid}", "/NH"],
                    capture_output=True, text=True, timeout=5,
                )
                return str(pid) in result.stdout
            except Exception:
                return False
        else:
            # Unix: signal 0 checks existence without affecting the process
            try:
                os.kill(pid, 0)
                return True
            except (OSError, ProcessLookupError):
                return False

    def terminate(self, pid: int) -> bool:
        """Terminate a process gracefully. Returns True if signal was sent."""
        if IS_WINDOWS:
            try:
                subprocess.run(
                    ["taskkill", "/PID", str(pid), "/F"],
                    capture_output=True, timeout=10,
                )
                return True
            except Exception:
                return False
        else:
            try:
                os.kill(pid, 15)  # SIGTERM
                return True
            except (OSError, ProcessLookupError):
                return False

    def open_url(self, url: str) -> None:
        """Open a URL in the default browser."""
        import webbrowser
        webbrowser.open(url)


# ─── Environment ───────────────────────────────────────────────────────

class _Env:
    """Cross-platform environment utilities."""

    def clean_env(self, *, passthrough: list[str] | None = None) -> dict[str, str]:
        """Build a clean subprocess environment.

        Strips CLAUDE* vars (nested session prevention), sets sensible defaults.
        Optional passthrough list for integration-specific vars.
        """
        home = paths.home()
        result = {
            "HOME": home,
            "USERPROFILE": home,  # Windows compat
            "USER": os.environ.get("USER", os.environ.get("USERNAME", "")),
            "PATH": paths.default_path(),
            "SHELL": paths.default_shell(),
            "LANG": os.environ.get("LANG", "en_US.UTF-8"),
            "TERM": "dumb",
            "HARNESS_ROOT": os.environ.get("HARNESS_ROOT", ""),
        }

        if IS_WINDOWS:
            # Windows needs these for subprocess spawning
            for var in ("SystemRoot", "COMSPEC", "TEMP", "TMP", "LOCALAPPDATA", "APPDATA"):
                val = os.environ.get(var)
                if val:
                    result[var] = val

        if passthrough:
            for var in passthrough:
                val = os.environ.get(var)
                if val:
                    result[var] = val

        # Remove empty values
        return {k: v for k, v in result.items() if v}

    def strip_claude_vars(self, base_env: dict[str, str] | None = None) -> dict[str, str]:
        """Return env dict with all CLAUDE* vars removed."""
        source = base_env if base_env is not None else dict(os.environ)
        return {k: v for k, v in source.items() if not k.startswith("CLAUDE")}


# ─── Scheduler Detection ──────────────────────────────────────────────

class _Scheduler:
    """Detect and report the platform scheduler."""

    def name(self) -> str:
        """Name of the platform scheduler."""
        if IS_MACOS:
            return "launchd"
        if IS_WINDOWS:
            return "task-scheduler"
        return "systemd"  # Linux default assumption

    def is_available(self) -> bool:
        """Check if the scheduler is usable."""
        if IS_MACOS:
            return shutil.which("launchctl") is not None
        if IS_WINDOWS:
            return shutil.which("schtasks") is not None
        return shutil.which("systemctl") is not None


# ─── Module-level singletons ──────────────────────────────────────────

paths = _Paths()
proc = _Proc()
env = _Env()
scheduler = _Scheduler()
